design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,198 @@
+   desired configuration at the time of provisioning (we make no
+   assumptions about preserving the state of the storage system or its
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
+
+For hardware RAID, the inputs describe how to combine physical disks
+into a RAID volume via the RAID controller hardware in the host.
+
+```yaml
+spec:
+  raid:
+    hardwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        name: volume-name
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
+
+For software RAID, the inputs describe how to configure the operating
+system to see physical disks in a RAID configuration through software
+drivers, without using any special hardware.
+
+```yaml
+spec:
+  raid:
+    softwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        physicalDisks:
+          - deviceName: /dev/sda
+          - deviceName: /dev/sdb
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The disks to use are specified using the same syntax as the
+  `rootDeviceHints`.
+
+Software RAID is constrained in other ways:
+
+- There must be 0, 1, or 2 software RAID devices.
+- The first software RAID volume must use RAID level 1 to ensure the
+  host boots in the event of a disk failure.
+- The second software RAID volume may use RAID level `0`, `1`, or
+  `1+0`.
+
+Regardless of the type of RAID used, the first volume defined is used
+as the root volume.
+
+The status fields of the `BareMetalHost` expand to include the RAID
+configuration used for the most recent provisioning operation.
+
+```yaml
+status:
+  provisioning:
+    raid:
+      hardwareVolumes: ...
+      softwareVolumes: ...
+```
+
+### Controller Changes
+
+Ironic applies RAID configuration during a cleaning phase, separate
```

Comment:
Has consideration been given to applying the RAID configuration during deployment? Ironic deploy steps facilitate that.

Reply:
Not in this PR, we could extend this implementation for deployment step based configuration as well, if we see sufficient interest in it.

Reply:
Perhaps I misunderstood the implementation. I thought the cleaning steps did happen before provisioning. If not, then I think we need to change the implementation so that's where the setup occurs.

Reply:
In the current implementation, the `cleaning steps` did before `provisioning`.

Reply:
> Perhaps I misunderstood the implementation. I thought the cleaning steps did happen before provisioning. If not, then I think we need to change the implementation so that's where the setup occurs.

Like @Hellcatlk has mentioned, It _is_ happening before provisioning. 

What @rpioso meant is, there are two mechanisms for delivering configurations to a node. One is as a cleaning step (performed earlier) and the other is a deployment step (performed later) in the "ironic node life-cycle". 

The latter can be added as an addendum to this later, if there is desire for it in the BMO. AFAIK, there is no objective preference between the two, it's a matter of choice.

Reply:
It sounds like the current implementation is OK, then.

Reply:
Yes, it is.

Reply:
The benefit of using deploy steps is that we don't need new states on the BMO side, everything happens during "Provisioning"

Reply:
I see. We can have this enhancement later then. Let's leave the implementation on ``clean_step`` for now.

Reply:
If we have a way to implement this without adding more states to the metal3 state machine, that seems like a better approach. Are there any down sides?

Reply:
I seem to vaguely recall that iDRAC RAID configuration had problems when used via a deploy step. I'll let @rpioso comment on that.

Also using deploy steps is a bit inconvenient right now (see https://opendev.org/openstack/ironic-tempest-plugin/src/branch/master/ironic_tempest_plugin/tests/scenario/baremetal_standalone_manager.py#L577-L595 for an example), but we're planning on fixing it in ironic (tracked in https://storyboard.openstack.org/#!/story/2008043).

Reply:
Currently, does `gophercloud` support `deploy step`? I did not find it in `gophercloud`.

Reply:
I doubt it would. But as @dtantsur says, the deploy steps could cause issues, so it's safe to stick to clean steps for now.